boilerplate/openshift/golang-osd-operator/csv-generate/common-generate-operator-bundle.py

```diff
@@ -0,0 +1,121 @@
+#    sys.exit(1)
+
+operator_name   = args.operator_name
+outdir          = args.output_dir
+prev_version    = args.previous_version
+git_num_commits = args.commit_number
+git_hash        = args.commit_hash
+operator_image  = args.operator_image
+
+full_version = "%s.%s-%s" % (VERSION_BASE, git_num_commits, git_hash)
+print("Generating CSV for version: %s" % full_version)
+
+if not os.path.exists(outdir):
+    os.mkdir(outdir)
+
+version_dir = os.path.join(outdir, full_version)
+if not os.path.exists(version_dir):
+    os.mkdir(version_dir)
+
+with open('config/templates/csv-template.yaml'.format(operator_name), 'r') as stream:
+    csv = yaml.load(stream)
+
+csv['spec']['customresourcedefinitions']['owned'] = []
+
+# Copy all CSV files over to the bundle output dir:
+# Copy all CRD files over to the bundle output dir:
+crd_files = [ f for f in os.listdir('deploy/crds') if f.endswith('_crd.yaml') ]
+for file_name in crd_files:
+    full_path = os.path.join('deploy/crds', file_name)
+    if (os.path.isfile(os.path.join('deploy/crds', file_name))):
+        shutil.copy(full_path, os.path.join(version_dir, file_name))
+    # Load CRD so we can use attributes from it
+    with open("deploy/crds/{}".format(file_name), "r") as stream:
+        crd = yaml.load(stream)
+    # Update CSV template customresourcedefinitions key
+    csv['spec']['customresourcedefinitions']['owned'].append(
+        {
+            "name": crd["metadata"]["name"],
+            "description": crd["spec"]["names"]["kind"],
+            "displayName": crd["spec"]["names"]["kind"],
+            "kind": crd["spec"]["names"]["kind"],
+            "version": crd["spec"]["version"]
+        }
+    )
+
+csv['spec']['install']['spec']['clusterPermissions'] = []
+
+# Add operator role to the CSV:
+with open('deploy/role.yaml', 'r') as stream:
+    operator_role = yaml.load(stream)
+    csv['spec']['install']['spec']['clusterPermissions'].append(
+        {
+            'rules': operator_role['rules'],
+            'serviceAccountName': operator_name,
+        })
+
+# Add our deployment spec for the hive operator:
+with open('deploy/operator.yaml', 'r') as stream:
+    operator_components = []
+    operator = yaml.load_all(stream)
+    for doc in operator:
+        operator_components.append(doc)
+    # There is only one yaml document in the operator deployment
+    operator_deployment = operator_components[0]
+    csv['spec']['install']['spec']['deployments'][0]['spec'] = operator_deployment['spec']
+
+# Update the deployment to use the defined image:
+csv['spec']['install']['spec']['deployments'][0]['spec']['template']['spec']['containers'][0]['image'] = operator_image
+
+# Update the versions to include git hash:
+csv['metadata']['name'] = "{}.v{}".format(operator_name, full_version)
+csv['spec']['version'] = full_version
+csv['spec']['replaces'] = "{}.v{}".format(operator_name, prev_version)
```

Comment:
We need a way to not write this when this is the first deploy as there will be no "replaces".

Reply:
This is what I did in dvo: https://github.com/app-sre/deployment-validation-operator/blob/master/hack/generate-operator-bundle-contents.py#L72-L73

But we have to be sure that the CSV template does not contain a `replaces` field

Reply:
This is where we would want it to be legal to run with `-p ""` or some other signal of "no previous version".

In that case, we could either make sure `replaces` is absent from the template and add it here if necessary, as you suggest; or we could just delete it here.

```suggestion
if prev_version:
    csv['spec']['replaces'] = "{}.v{}".format(operator_name, prev_version)
else:
    del csv['spec']['replaces']
```

Or more carefully (works even if `replaces` isn't in the template):

```suggestion
if prev_version:
    csv['spec']['replaces'] = "{}.v{}".format(operator_name, prev_version)
else:
    csv.pop(['spec']['replaces'], None)
```